src/compiler/emitter.ts

```diff
@@ -6105,115 +6112,83 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
-                                    : (<ImportEqualsDeclaration>importNode).name;
-
-                            if (defaultName) {
-                                // emit re-export for imported default name
-                                // import n1 from 'foo1'
-                                // import n2 = require('foo2')
-                                // export {n1}
-                                // export {n2}
-                                emitExportMemberAssignments(defaultName);
                                 writeLine();
-                            }
-
-                            if (importNode.kind === SyntaxKind.ImportDeclaration &&
-                                (<ImportDeclaration>importNode).importClause.namedBindings) {
-
-                                let namedBindings = (<ImportDeclaration>importNode).importClause.namedBindings;
-                                if (namedBindings.kind === SyntaxKind.NamespaceImport) {
-                                    // emit re-export for namespace
-                                    // import * as n from 'foo'
-                                    // export {n}
-                                    emitExportMemberAssignments((<NamespaceImport>namedBindings).name);
+                                // save import into the local
+                                write(`${importVariableName} = ${parameterName};`);
+                                writeLine();
+                                break;
+                            case SyntaxKind.ExportDeclaration:
+                                Debug.assert(importVariableName !== "");
+
+                                if ((<ExportDeclaration>entry).exportClause) {
+                                    // export {a, b as c} from 'foo'
+                                    // emit as:
+                                    // var reexports = {}
+                                    // reexports['a'] = _foo["a"];
+                                    // reexports['c'] = _foo["b"];
+                                    // exports_(reexports);
+                                    let reexportsVariableName = makeUniqueName("reexports");
                                     writeLine();
-                                }
-                                else {
-                                    // emit re-exports for named imports
-                                    // import {a, b} from 'foo'
-                                    // export {a, b as c}
-                                    for (let element of (<NamedImports>namedBindings).elements) {
-                                        emitExportMemberAssignments(element.name || element.propertyName);
+                                    write(`var ${reexportsVariableName} = {};`);
+                                    writeLine();
+                                    for (let e of (<ExportDeclaration>entry).exportClause.elements) {

```

Comment:
i would replace this with:
```ts
for (let element of (<NamedImports>namedBindings).elements) {
    emitExportMemberAssignments(element.name || element.propertyName);
    //write(`var ${reexportsVariableName} = {};`);
    write(`${exportFunctionForFile}({`);
    increaseIndent();
    writeLine();
    for (let e of (<ExportDeclaration>entry).exportClause.elements) {
        write(`"`);
        emitNodeWithoutSourceMap(e.name);
        write(`": ${parameterName}["`);
        emitNodeWithoutSourceMap(e.propertyName || e.name);
        write(`"];`);
        writeLine();
    }
    decreaseIndent();
    write(`});`);
}
```
saves you the extra variable declaration and one line less :D